src/core/Merger.cpp

```diff
@@ -0,0 +1,520 @@
+
+void Merger::eraseEntry(Entry* entry)
+{
+    Database* database = entry->database();
+    // most simple method to remove an item from DeletedObjects :(
+    const QList<DeletedObject> deletions = database->deletedObjects();
+    Group* parentGroup = entry->group();
+    const bool groupUpdateTimeInfo = parentGroup ? parentGroup->canUpdateTimeinfo() : false;
+    if (parentGroup) {
+        parentGroup->setUpdateTimeinfo(false);
+    }
+    delete entry;
+    if (parentGroup) {
+        parentGroup->setUpdateTimeinfo(groupUpdateTimeInfo);
+    }
+    database->setDeletedObjects(deletions);
+}
+
+void Merger::eraseGroup(Group* group)
+{
+    Database* database = group->database();
+    // most simple method to remove an item from DeletedObjects :(
+    const QList<DeletedObject> deletions = database->deletedObjects();
+    Group* parentGroup = group->parentGroup();
+    const bool groupUpdateTimeInfo = parentGroup ? parentGroup->canUpdateTimeinfo() : false;
+    if (parentGroup) {
+        parentGroup->setUpdateTimeinfo(false);
+    }
+    delete group;
+    if (parentGroup) {
+        parentGroup->setUpdateTimeinfo(groupUpdateTimeInfo);
+    }
+    database->setDeletedObjects(deletions);
+}
+
+Merger::ChangeList
+Merger::resolveEntryConflict(const MergeContext& context, const Entry* sourceEntry, Entry* targetEntry)
+{
+    ChangeList changes;
+    // We need to cut off the milliseconds since the persistent format only supports times down to seconds
+    // so when we import data from a remote source, it may represent the (or even some msec newer) data
+    // which may be discarded due to higher runtime precision
+    const auto timeTarget = Clock::serialized(targetEntry->timeInfo().lastModificationTime());
+    const auto timeSource = Clock::serialized(sourceEntry->timeInfo().lastModificationTime());
+
+    Group::MergeMode mergeMode = m_mode == ModeDefault ? context.m_targetGroup->mergeMode() : m_mode;
+
+    switch (mergeMode) {
+    case Group::KeepBoth:
+        // if one entry is newer, create a clone and add it to the group
+        if (timeTarget > timeSource) {
+            Entry* clonedEntry = sourceEntry->clone(Entry::CloneNewUuid | Entry::CloneIncludeHistory);
+            moveEntry(clonedEntry, context.m_targetGroup);
+            markOlderEntry(clonedEntry);
+            changes << tr("Adding backup for older source %1 [%2]")
+                           .arg(sourceEntry->title())
+                           .arg(sourceEntry->uuid().toHex());
+        } else if (timeTarget < timeSource) {
+            Entry* clonedEntry = sourceEntry->clone(Entry::CloneNewUuid | Entry::CloneIncludeHistory);
+            moveEntry(clonedEntry, context.m_targetGroup);
+            markOlderEntry(targetEntry);
+            changes << tr("Adding backup for older target %1 [%2]")
+                           .arg(targetEntry->title())
+                           .arg(targetEntry->uuid().toHex());
+        }
+        break;
+
+    case Group::KeepNewer:
```

Comment:
How do you mean "transaction" in this context? Currently, there is no generalized transaction concept in KeePassXC as far as I know. I think, such a concept would be good and may improve the merging and sharing feature, but for me it seems out of scope for now. Maybe there should be a generalized feature ticket requesting for a consistent transaction concept.

Reply:
We do not have a transaction mechanism with rollback, but a capture mechanism for creating history snapshots with beginUpdate() and endUpdate(). That's what I mean.

Reply:
The capture mechanism is just available for `Entry` since `Group` does not feature a history. Furthermore a merge should never introduce new data during the operation (except it is forced by `KeepBoth`).

Reply:
Yeah, Groups do not, but for Entries we should create a history item. Otherwise KeepNewer or KeepOlder may inadvertently overwrite data.

Reply:
Up to now, I thought `KeepNewer` and `KeepExisting` were both intentionally written that way, so they would lose the discarded data - but this should be addressed separately (maybe as a bug for the current release?)
The `beginUpdate()` `endUpdate()` bracket should not be used for history merge, since the older of both entries is pushed to history as latest history item. In fact, using them would again change timestamps within the database which would lead to unintended behavior.

Reply:
I'm not saying that our current merge strategy is flawless, but we have the chance to improve it here. No database operation should ever discard data without backing them up to a history item first. Actually, you don't really have to use `beginUpdate()` and `endUpdate()`, since you are not introducing edits which haven't been captured before, but I would suggest the following modified merging strategy for KeepNewer and KeepOlder:

- merge histories (they should be almost identical in most cases)
- if the strategy is KeepNewer, add older entry to history and keep new
- if the strategy is KeepOlder, add new entry to history and keep old one

It's basically the same as Synchronize, but you can choose which Entry to keep at the top. I acknowledge that hard-replacing Entries does more strictly what KeepNewer and KeepOlder imply, but I'm not comfortable with ever modifying an entry without making the modification reversible.

Reply:
Okay, I'll extend those merge methods accordingly. In case of `KeepExisting` (which would be keep own in my opinion) and the existing one is older, the merge strategy would look like this (from old to new):
* old history
* own entry
* new entry
* copy of own entry with merge timestamp as modification time
This way, the history should have the correct timeline (as long it is large enough) and the user has the requested entry on top. We may need to think about merging of deleted objects again - currently they disregard the merge strategy, which may not be the correct behavior.

Reply:
Thinking about the request, it seems to me, that we want to keep the history of both entries regardless of the merge strategy which would result `Synchronize` strategy for `KeepNewer` and the `Synchronize` strategy for `KeepExisting` but with duplicating the old entry and putting it on top. Am i right?

Reply:
Sounds reasonable, yes.